[flutter_tools] Display "no platforms" message based on results when creating plugins project

cyanglaz opened the pull request:
## Description

1. This PR mainly focused displaying the "no platforms" message based on the plugin project's file structure.
It also moved the "no platforms" message to the very end of the log to make it more visible when executing `flutter create -t plugin .` The "no platforms" message was also edited to reflect the change.

2. While I'm here, I also moved the "update pubspec" message towards the end a bit, also edited the final message of creating plugins to encourage users to find the "update pubspec" message. (Although this could be improved to be more smart, but it requires more refactoring of the code so I left it there for now)

3. Cleaned up some stale TODO comments

## Related Issues

Fixes https://github.com/flutter/flutter/issues/66058
Fixes https://github.com/flutter/flutter/issues/70492

## Tests

I added the following tests:

created plugin with no --platforms flag should not print `no platforms` message if the existing plugin supports a platform.
created plugin supports no platforms should print `no platforms` message

Also edited some existing test to also check for this message.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Commit:
move no platform views message and eidt pubspec message to the end of the log

cyanglaz commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -615,6 +629,27 @@ To edit platform code in an IDE see https://flutter.dev/developing-packages/#edi
       );
     }
 
+    final FlutterProject project = FlutterProject.fromDirectory(directory);
+    final bool generateAndroid = templateContext['android'] == true;
+    if (generateAndroid) {
+      gradle.updateLocalProperties(
+        project: project, requireAndroidSdk: false);
+    }
+
+    final String projectName = templateContext['projectName'] as String;
+    final String organization = templateContext['organization'] as String;
+    final String androidPluginIdentifier = templateContext['androidIdentifier'] as String;
+    final String exampleProjectName = projectName + '_example';
+    templateContext['projectName'] = exampleProjectName;
+    templateContext['androidIdentifier'] = _createAndroidIdentifier(organization, exampleProjectName);
+    templateContext['iosIdentifier'] = _createUTIIdentifier(organization, exampleProjectName);
+    templateContext['macosIdentifier'] = _createUTIIdentifier(organization, exampleProjectName);
+    templateContext['description'] = 'Demonstrates how to use the $projectName plugin.';
+    templateContext['pluginProjectName'] = projectName;
+    templateContext['androidPluginIdentifier'] = androidPluginIdentifier;
+
+    generatedCount += await _generateApp(project.example.directory, templateContext, overwrite: overwrite, pluginExampleApp: true);
+
```
This diff looks weird. I basically moved the below `if (addPlatformsToExistingPlugin) {` block from before this code to after this code.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -39,7 +39,12 @@ const List<String> _kAvailablePlatforms = <String>[
 
 const String _kNoPlatformsErrorMessage = '''
-The plugin project was generated without specifying the `--platforms` flag, no new platforms are added.
+The plugin project does not seem to support any platform.
```
nit: "seem" is kind of a weak statement. I would say something stronger like:

"The plugin project is not configured to support any platforms"

cyanglaz replied:
I made it weak due to the "web" issue I mentioned below. I'm not sure what I prefer, do you think we should still make it a strong statement then mentioned the "web" issue too?

jonahwilliams commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -39,7 +39,12 @@ const List<String> _kAvailablePlatforms = <String>[
-The plugin project was generated without specifying the `--platforms` flag, no new platforms are added.
+The plugin project does not seem to support any platform.
+
+This determination is based on the folders and files in the project directory assuming best practise of naming
```
I'm not sure what this means - I don't think we support configuring the platform folders

cyanglaz replied:
This message is mainly for web.
Because there is no "web" folder, the only way we determine if the plugin supports web is to check if `lib/plugin_web.dart` exists.
However, people can edit that file to a different name. So we have to say something about it like "this claim is not deterministic if the plugin already supports web"

Any suggestion on the message? :) 

How about:

If the plugin currently only supports web, you might see this message because the web file name was not ending with "web.dart". You can discard this message if that was the case.

jonahwilliams replied:
but the web still needs a manifest entry

jmagman replied:
I think you should reach out to the UXR team for help with this one.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -500,6 +505,10 @@ Your plugin code is in $relativePluginMain.
 To edit platform code in an IDE see https://flutter.dev/developing-packages/#edit-plugin-package.
+
+If you are adding new platforms to an existing plugin, you might need to update the pubspec.yaml file.
+Scroll up to find a detailed instruction on how to update the pubspec.yaml.
```
> Scroll up to find a detailed instruction on how to update the pubspec.yaml

Maybe something more specific like "at the start of this message?"

If they didn't see the message before, perhaps we should link to some flutter.dev documentation on this topic instead

jmagman commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -39,7 +39,12 @@ const List<String> _kAvailablePlatforms = <String>[
 
 const String _kNoPlatformsErrorMessage = '''
-The plugin project was generated without specifying the `--platforms` flag, no new platforms are added.
+The plugin project does not seem to support any platform.
```
"does not" instead of "does not seem"

jmagman commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -39,7 +39,12 @@ const List<String> _kAvailablePlatforms = <String>[
-The plugin project was generated without specifying the `--platforms` flag, no new platforms are added.
+The plugin project does not seem to support any platform.
+
+This determination is based on the folders and files in the project directory assuming best practise of naming
```
practice

jmagman reviewed (commented):
This whole thing is getting wordy.  Can you reach out to @InMatrix and team to tighten this whole flow up?  The wall of text is usually not a good way to communicate "what went wrong, how do I fix it"

cyanglaz reviewed (commented):
@jmagman @jonahwilliams 

Updated the message after offline discussion with @InMatrix 

Also updated the code so that the "edit pubpsec.yaml" message is also displayed at the end of the log, so we don't need the "scroll up to see pubspec.yaml templates.." message anymore.

PTAL :)

jmagman commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -200,6 +204,17 @@ class FlutterManifest {
   }
 
+  /// Like [supportedPlatforms], but only returns the valid platforms that are supported in flutter plugins.
+  Map<String, dynamic> get validSupportedPlatforms {
```
What if we moved this logic into `_validateMultiPlatformYaml` so it fails validation if you add an unsupported platform, instead of silently dropping it on the ground?

cyanglaz replied:
When the plugin is created with no `--platforms` flag, we explicitily added an invalid platform to make the pubspec format valid. 
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/plugin/pubspec.yaml.tmpl#L40
If we throw in `_validateMultiPlatformYaml`, it would break.

jmagman reviewed (commented):
Error message looking better!  Would you mind adding a screen shot of what it looks like now?

cyanglaz commented:
@jmagman Screenshots:
<img width="970" alt="Screen Shot 2020-11-11 at 4 55 57 PM" src="https://user-images.githubusercontent.com/3756895/98881432-d381bf00-243e-11eb-8719-b3f53ebab2b1.png">

![Uploading Screen Shot 2020-11-11 at 4.57.54 PM.png…]()

cyanglaz commented:
Screenshots for adding platforms:
<img width="1059" alt="Screen Shot 2020-11-11 at 4 57 54 PM" src="https://user-images.githubusercontent.com/3756895/98881917-f2348580-243f-11eb-9688-486c2b6124d8.png">

jonahwilliams commented:
I think we still need to cut down the amount of output here, it's a bit overwhelming. Some ideas

1. Don't run doctor when adding a plugin, its just more noise
2. Create website documentation and link to that for caveats, keep the documentation to the happy path. I don't think that caveat about platforms: is that important, just update the snippet to include the platform key too and I think people will figure it out.
3. Pending some feedback from UXR, I would make the actionable text more even more obvious and put it at the very end:

"""
Action Required: One sentence thing to do

Link for more info:
"""

cyanglaz commented:
> 1. Don't run doctor when adding a plugin, its just more noise
@jonahwilliams Did you mean "when adding a platform"?

cyanglaz commented:
> 2. Create website documentation and link to that for caveats, keep the documentation to the happy path. I don't think that caveat about platforms: is that important, just update the snippet to include the platform key too and I think people will figure it out.

**For `flutter create -t plugin .`**

```
Action Required: Must specify at least one platform using --platforms.

Link for more info: https://flutter.dev/docs/development/packages-and-plugins/developing-packages#step-1-create-the-package-1
```

**For adding platforms in existing plugin using `flutter create -t plugin --platforms=android`**


```
Action Required: The `pubspec.yaml` under the project directory must be updated to support android.

Link for more info: https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms
```

@jonahwilliams @InMatrix @jmagman WDYT?

jmagman reviewed (approved):
This test is failing:
```
04:04 +1267 ~3 -1: test/general.shard/flutter_manifest_test.dart: FlutterManifest validSupportedPlatforms return null if the platform keys are not valid [E]                                           
  Expected: <null>
    Actual: {}
  
  package:test_api                                     expect
  test/general.shard/flutter_manifest_test.dart 991:5  main.<fn>
  dart:async                                           runZoned
  test/src/common.dart 222:14                          testWithoutContext.<fn>
  test/src/common.dart 221:18                          testWithoutContext.<fn>
  test/src/common.dart 192:18                          test.<fn>
  test/src/common.dart 188:5                           test.<fn>
```

Other than that, I'm fine with this change, I think it's an improvement on what we have now.  And https://github.com/flutter/flutter/pull/70475 should make it less noisy.  We can open another issue to improve the message if @jonahwilliams has suggestions.

LGTM assuming the test gets fixed.

Commit:
remove unused var

Commit:
update output

Commit:
cleanup

jonahwilliams commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -494,3 +451,45 @@ https://flutter.dev/docs/development/packages-and-plugins/developing-packages#pl
+Must specify at least one platform using --platforms. See the following for details:
+https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin
+
+''';
```
```suggestion
''';

```

jonahwilliams commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -494,3 +451,45 @@ https://flutter.dev/docs/development/packages-and-plugins/developing-packages#pl
+
+void _printPluginDirectoryLocationMessage(String pluginPath, String projectName, String platformsString) {
+  final String relativePluginMain = globals.fs.path.join(pluginPath, 'lib', '$projectName.dart');
+final String relativeExampleMain = globals.fs.path.join(pluginPath, 'example', 'lib', 'main.dart');
```
```suggestion
  final String relativeExampleMain = globals.fs.path.join(pluginPath, 'example', 'lib', 'main.dart');
```

jonahwilliams commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -494,3 +451,45 @@ https://flutter.dev/docs/development/packages-and-plugins/developing-packages#pl
+List<String> _getSupportedPlatformsInPlugin(Directory projectDir) {
+  final String pubspecPath = globals.fs.path.join(projectDir.absolute.path, 'pubspec.yaml');
+  final FlutterManifest manifest = FlutterManifest.createFromPath(pubspecPath, fileSystem: globals.fs, logger: globals.logger);
+  final List<String> platforms = manifest.validSupportedPlatforms == null
+                                ? <String>[]
+                                : manifest.validSupportedPlatforms.keys.toList();
+  return platforms;
```
```suggestion
  final List<String> platforms = manifest.validSupportedPlatforms == null
    ? <String>[]
    : manifest.validSupportedPlatforms.keys.toList();
  return platforms;
```

Commit:
Update packages/flutter_tools/lib/src/commands/create.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

Commit:
Update packages/flutter_tools/lib/src/commands/create.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

Commit:
fix tests

Commit:
Merge branch 'flutter_create' of github.com:cyanglaz/flutter into flutter_create

Commit:
small code refactor

cyanglaz commented:
@jmagman @jonahwilliams I updated the code to reflect what we have in the discussion doc. I think the message looks pretty concise and clear now. 
I also updated the link to use a go link that will be available once https://github.com/flutter/website/pull/4997 lands.

PTAL

Commit:
update output

Commit:
ci

Commit:
only print out platform directoy information when platforms are requested

Commit:
add extra line

cyanglaz commented:
Here are screenshots, 1st is no platforms, 2nd is add platforms
<img width="1086" alt="Screen Shot 2020-11-22 at 9 48 52 AM" src="https://user-images.githubusercontent.com/3756895/99911318-6248d400-2ca8-11eb-93b7-1ddf3a7eb211.png">

<img width="1081" alt="Screen Shot 2020-11-22 at 9 50 06 AM" src="https://user-images.githubusercontent.com/3756895/99911307-4ba27d00-2ca8-11eb-89bd-cb2fc63327c5.png">

jonahwilliams reviewed (approved):
LGTM

Thanks for kicking this one around a bit more

cyanglaz commented:
Will land this after https://github.com/flutter/website/pull/4997 lands
